### PR TITLE
feat(xmlupload): check if the encoding in the text-prop is consistent (DEV-3296)

### PR DESCRIPTION
**Pull request by Nora-Olivia-Ammann**

### CODE REVIEW

File: src/dsp_tools/utils/xml_validation_models.py
Comment (jnussbaum) [inline src/dsp_tools/utils/xml_validation_models.py:117]: The order how you do things in this method seems very unnatural to me. You do:

- compose a base message
- create a dataframe
- if the dataframe is big enough to be saved --> return it together with the message
    - on the caller side, you add the info that the file was saved. For me, this would belong here.
- compose an additional message from the dataframe 
    - here it gets really confusing: at that point, the dataframe will not be used, but you use it to create an "additional" message 
    - it is not clear what "additional" means in this context
- return the additional message.

My suggestion for a more natural flow of things: 

- compose a base message
- based on `problematic_resources`, determine whether to save a df or not
- if not:
    - based on `problematic_resources`, compose a print message
    - return it
- if yes:
    - create a dataframe
    - to the base message, add that a dataframe was saved, and where it was saved
    - return df + message

File: src/dsp_tools/utils/xml_validation_models.py
Comment (jnussbaum) [inline src/dsp_tools/utils/xml_validation_models.py:128]:
```suggestion
                "Resource ID": [x.resource_id for x in self.problematic_resources],
                "Property Name": [x.property_name for x in self.problematic_resources],
```

File: src/dsp_tools/utils/xml_validation_models.py
Comment (jnussbaum) [inline src/dsp_tools/utils/xml_validation_models.py:137]: it seems unnatural to me that these 2 methods are static. I would expect them both to be normal instance methods, and to call them with self.xy()

File: test/unittests/utils/test_xml_validation_models.py
Comment (jnussbaum) [inline test/unittests/utils/test_xml_validation_models.py:35]: if you follow my recommendation to return an empty list instead of None, then you have to write:

```suggestion
    assert not res
```

File: test/unittests/utils/test_xml_validation_models.py
Comment (jnussbaum) [inline test/unittests/utils/test_xml_validation_models.py:38]: why mixed?

File: test/unittests/utils/test_xml_validation_models.py
Comment (jnussbaum) [inline test/unittests/utils/test_xml_validation_models.py:193]:
```suggestion
                <resource label="resD" restype=":TestThing2" id="resD" permissions="res-default">
```

File: test/unittests/utils/test_xml_validation_models.py
Comment (jnussbaum) [inline test/unittests/utils/test_xml_validation_models.py:267]:
```suggestion
                <resource label="resD" restype=":TestThing2" id="resD" permissions="res-default">
```

File: test/unittests/utils/test_xml_validation_models.py
Comment (jnussbaum) [inline test/unittests/utils/test_xml_validation_models.py:307]:
```suggestion
                <resource label="resD" restype=":TestThing2" id="resD" permissions="res-default">
```

**Review by jnussbaum — CHANGES_REQUESTED** [review]
I like the intention of the PR, and also most of how it is done, but there are 2 main reasons why I'd like to ask you to revise it:

- the present PR contains some changes from another pending PR
- there are many parts of the code that could be simplified / made more pythonic (see my comments)

File: src/dsp_tools/utils/xml_validation_models.py
Comment (Nora-Olivia-Ammann) [reply src/dsp_tools/utils/xml_validation_models.py:89]: I guess since the XML does not allow any other values we can scrap it.

File: src/dsp_tools/utils/xml_validation_models.py
Comment (Nora-Olivia-Ammann) [reply src/dsp_tools/utils/xml_validation_models.py:137]: They are static because they do not need to have access to any class variables at any point. In case of static functions I prefer to declare them so, because one it is correct and for another it makes testing simpler because you need not create a class to call the function.

File: src/dsp_tools/utils/xml_validation_models.py
Comment (Nora-Olivia-Ammann) [reply src/dsp_tools/utils/xml_validation_models.py:1]: I am not entirely happy with this either. My plan for this file is that it will also host the check if the correct encoding with respect to the property is used. It seemed a bit like overkill to create two files, one with the check functions and another with the model, but I am open for suggestions.

File: src/dsp_tools/utils/xml_validation_models.py
Comment (Nora-Olivia-Ammann) [reply src/dsp_tools/utils/xml_validation_models.py:10]: I disagree:
- In other files that generate a error string (eg. src/dsp_tools/commands/xmlupload/models/ontology_problem_models.py) These are listed at the top, making it consistent.
- The idea is that the messages will be uniform, meaning other problems should have access to it. I would be happy to look into making these constants for the entire repo, but that should be another PR.
- I intend to add the problem model that generates the error string if the wrong encoding is used with respect to the property in this file. This will also need access to these variables.

File: src/dsp_tools/utils/xml_validation_models.py
Comment (Nora-Olivia-Ammann) [reply src/dsp_tools/utils/xml_validation_models.py:104]: Where the df is saved:
- I did it this way because in the past that is how we structured it. The rationale if I remember correctly is for better testing.
- Another reason for the location of the save message is, that it is in fact not saved in that function, therefore the message would be technically wrong in my opinion.

Composing the df:
- The reason why, I did it first, is because for a clean message, I want to group by resource id and then sort by prop name. This can both be done nicely in a df, since we may in any case create a df, I did not see a reason not to do it this way. In my opinion any other method I can think of that creates the same result is more complicated, so I don't think we should change this.

Additional msg:
- I will restructure as suggested below, then the problem with the variable name is solved.

File: src/dsp_tools/utils/xml_validation_models.py
Comment (jnussbaum) [reply src/dsp_tools/utils/xml_validation_models.py]: Is it possible that you encountered this with pandas dataframe comparison? I'm almost 100% sure that in pure python, my suggestion is cleaner and will never raise a future warning. I don't think that your variant is a good habit. (Except for pandas, there it might be necessary)

File: src/dsp_tools/utils/xml_validation_models.py
Comment (jnussbaum) [reply src/dsp_tools/utils/xml_validation_models.py:10]: That sounds like a perfect use case for inheritance: If `Problem` was a class (not a protocol), and the `InconsistentTextValueEncodings` inherits from it, then you can add these fields to the base class, and all children will have the same values.

With the present solution, we have multiple files, each defines its own values, and if we want to change the values, we have to modify multiple files. 

So for right now, you can leave it as it is, but for the future, we should check if inheritance would be a better solution.

File: src/dsp_tools/utils/xml_validation_models.py
Comment (jnussbaum) [reply src/dsp_tools/utils/xml_validation_models.py:122]: Even if these methods don't access class variables, they access (static) class methods. And for this, I think that static is less elegant than non-static. For example, I think that on line 137, `InconsistentTextValueEncodings._make_msg_for_one_resource()` is slightly more difficult to understand than `self._make_msg_for_one_resource()`.

But perhaps it's a matter of taste, so if you prefer them to be static, it's okay for me.

File: src/dsp_tools/utils/xml_validation_models.py
Comment (jnussbaum) [reply src/dsp_tools/utils/xml_validation_models.py:1]: I would prefer to have to separate files, one called "models", one called "functions". But feel free to leave it as it is.

File: src/dsp_tools/utils/xml_validation.py
Comment (jnussbaum) [reply src/dsp_tools/utils/xml_validation.py:75]: I feel you, and I have done such things in the past myself, and I had to make the painful experience that mostly, you don't do yourself a favor. Because if the other PR is modified (as it is the case now), you have to migrate the  modifications. Plus, for the reviewer, it can be really confusing, because multiple things are mixed together.

But it's fine for now. I have approved the other PR. As soon as you have merged it and "tidied up" this one, I can give you an approve.

File: src/dsp_tools/utils/xml_validation_models.py
Comment (BalduinLandolt) [reply src/dsp_tools/utils/xml_validation_models.py]: In lxml you may also get these warnings, I think. But it's true that in pure python this behaviour guaranteed to work - the only case where you need `is not None` is when you want to distinguish actual `None` from some falsy value like empty string, empty list, `False`, etc.

File: src/dsp_tools/utils/xml_validation_models.py
Comment (Nora-Olivia-Ammann) [reply src/dsp_tools/utils/xml_validation_models.py:10]: I added it as a point to the notion page that discusses the config file.
https://www.notion.so/dasch-swiss/New-global-config-object-be2645e1e9a14e5da185dd8b3c6abd8f

File: src/dsp_tools/utils/xml_validation_models.py
Comment (Nora-Olivia-Ammann) [reply src/dsp_tools/utils/xml_validation_models.py:122]: Thanks for the suggestion, I think I prefer Balduin's suggestion.

File: src/dsp_tools/utils/xml_validation.py
Comment (Nora-Olivia-Ammann) [reply src/dsp_tools/utils/xml_validation.py]: No worries, I tend not to do that either but as you both are not officially working, I did not wait for an approve. I do appreciate that you take a look even though you are ill. Otherwise the mess would be even greater.

**Review by jnussbaum — APPROVED** [review]
LGTM

File: src/dsp_tools/utils/xml_validation.py
Comment (jnussbaum) [inline src/dsp_tools/utils/xml_validation.py:171]: That's outdated now.

File: src/dsp_tools/utils/xml_validation.py
Comment (jnussbaum) [inline src/dsp_tools/utils/xml_validation.py:197]: I have troubles understanding the function name. And when I compare it to what the function actually does, I have even more question marks. 

If I look at the code, I would say something like: 

- "find illegal text props"
- "find text props that use multiple encodings"

**Pull request merged**
